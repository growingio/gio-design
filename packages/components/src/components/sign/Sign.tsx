import React, { Children } from 'react';
import classnames from 'classnames';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import { ISignProps } from './interface';
import Number from './Number';
import Dot from './Dot';

export { TPlacement, TSignVariant } from './interface';

const toPascal = (hump: string) => hump.replace(/([A-Z])/g, '-$1').toLowerCase();

const Sign: React.FC<ISignProps> = ({
  variant = 'number',
  className,
  prefixCls: customPrefixCls,
  style,
  placement = 'rightTop',
  visible = true,
  count = 0,
  showZero = false,
  magnitude = 100,
  status = 'default',
  size = 'middle',
  children,
}: ISignProps) => {
  const prefix = usePrefixCls('sign', customPrefixCls);
  const noChildren = Children.count(children) === 0;

  const cls = classnames(className, prefix, {
    [`${prefix}--no-children`]: noChildren,
  });

  const supCls = classnames(`${prefix}__${variant}`, {
    [`${prefix}--${toPascal(placement)}`]: !noChildren,
  });

  return (
    <span className={cls}>
      {variant === 'dot' && (
        <Dot prefixCls={prefix} className={supCls} status={status} size={size} style={style} visible={visible} />
      )}
      {variant === 'number' && (
        <Number
          prefixCls={prefix}
          className={supCls}
          style={style}
          count={count}
          showZero={showZero}
          magnitude={magnitude}
          visible={visible}
        />
      )}
      {children}
    </span>
  );
};

export default Sign;
