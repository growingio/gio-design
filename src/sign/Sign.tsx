import React, { Children } from 'react';
import classnames from 'classnames';
import { usePrefixCls } from '@gio-design/utils';
import { ISignNumberProps } from './interface';
import SignNumber from './SignNumber';

export { TPlacement } from './interface';

const toPascal = (hump: string) => hump.replace(/([A-Z])/g, '-$1').toLowerCase();

const Sign: React.FC<ISignNumberProps> = ({
  className,
  prefixCls: customPrefixCls,
  style,
  placement = 'rightTop',
  visible = true,
  count = 0,
  showZero = false,
  magnitude = 100,
  children,
  offset,
}: ISignNumberProps) => {
  const prefix = usePrefixCls('sign-new', customPrefixCls);
  const noChildren = Children.count(children) === 0;

  const cls = classnames(className, prefix, `${prefix}__number-wrapper`, {
    [`${prefix}--no-children`]: noChildren,
  });

  const supCls = classnames(`${prefix}__number`, {
    [`${prefix}--${toPascal(placement)}`]: !noChildren,
  });

  return (
    <span className={cls}>
      <SignNumber
        prefixCls={prefix}
        className={supCls}
        style={style}
        count={count}
        showZero={showZero}
        magnitude={magnitude}
        visible={visible}
        offset={offset}
        placement={placement}
      />
      {children}
    </span>
  );
};

export default Sign;
