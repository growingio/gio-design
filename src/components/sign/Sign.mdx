import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Sign from './Sign';

# Sign 标记

<Subtitle>图标右上角的圆形徽标数字。</Subtitle>

## 使用规范

1. 导航中的某一项目的右上角。
2. Tab 中的某一项目的右上角。
3. 头像的右上角。
4. 通知图标的右上角。

## 参数说明

<ArgsTable of={Sign} />

## 代码演示

### 默认-数字标记

1. 一般出现在通知图标或头像的右上角，用于显示需要处理的消息条数，通过醒目视觉形式吸引用户处理。
2. 最大显示数字标识可选「一位数」或「两位数」，「一位数」最多显示 9，超过显示 9+；「两位数」最多显示 99，超过显示 99+。
3. 当数值为「0」时，不显示标记控件。
4. 点击查看后，数字标识消失。

[Example](/?path=/story/basic-components-sign--default)

### 状态标记

1. 用不同的颜色标记状态，绿色表示正常、黄色表示警告、红色表示错误、灰色表示不可用。一般使用在列表中，如：「数据中心-事件」、「组织管理-成员管理」等。
2. 用红色小圆点表示更新状态，点击查看后，小圆点消失。

[Example](/?path=/story/basic-components-sign--dot-sign)
