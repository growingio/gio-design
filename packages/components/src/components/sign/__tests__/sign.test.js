import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Sign from '..';

describe('Testing Sign', () => {
  it('should mount and unmount Sign with no error.', () => {
    const wrapper = mount(<Sign count={12} />);
    expect(() => {
      wrapper.setProps({});
      wrapper.unmount();
    }).not.toThrow();
  });

  it('should render number sign correctly', () => {
    const domTree = renderer
      .create(
        <Sign className='gio-customized-sign' count={100} magnitude={100}>
          <span>Number</span>
        </Sign>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  it('should render dot sign correctly', () => {
    const domTree = renderer
      .create(
        <Sign className='gio-customized-sign' variant='dot'>
          <span>Dot</span>
        </Sign>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  it('should hide.', () => {
    const wrapper = mount(
      <Sign className='gio-customized-sign' variant='dot' visible={false}>
        <span>Dot</span>
      </Sign>
    );

    expect(wrapper.find('.gio-sign__dot').at(0).hasClass('gio-sign--hide')).toBe(true);
  });

  it('should has error class.', () => {
    const wrapper = mount(
      <Sign className='gio-customized-sign' variant='dot' status='error'>
        <span>Dot</span>
      </Sign>
    );

    expect(wrapper.find('.gio-sign__dot').at(0).hasClass('gio-sign__dot--error')).toBe(true);
  });
});
