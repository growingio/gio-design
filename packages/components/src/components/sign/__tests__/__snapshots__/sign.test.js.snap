// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Sign should render dot sign correctly 1`] = `
<span
  className="gio-customized-sign gio-sign"
>
  <span
    className="gio-sign__dot gio-sign--right-top gio-sign__dot--default gio-sign__dot--middle"
  />
  <span>
    Dot
  </span>
</span>
`;

exports[`Testing Sign should render number sign correctly 1`] = `
<span
  className="gio-customized-sign gio-sign"
>
  <span
    className="gio-sign__number gio-sign--right-top"
  >
    99+
  </span>
  <span>
    Number
  </span>
</span>
`;
